### PR TITLE
Chore!: Improve the data model and performance of the state sync

### PR DESCRIPTION
**Pull request by izeigerman**
This update contains the following performance improvements to the state sync:
- Add dev intervals in bulk for new forward-only snapshots. Previously they were added one interval at a time.
- Update the `updated_ts` property using the `UPDATE` statement for all impacted snapshots at once when demoting snapshots from an environment.
- Only fetch the `plan_id` attribute and only update the `finalized_ts` attribute when finalizing an environment. Previously we serialized / deserialized the full environment object to perform this operation.
- Use the `UPDATE` statement to update the `unpaused_ts` attribute for all impacted snapshots at once when unpausing.

### CODE REVIEW

File: sqlmesh/core/state_sync/base.py [hunk 1]
```diff
@@ -417,11 +399,11 @@ def rollback(self) -> None:
         """Rollback to previous backed up state."""
 
     @abc.abstractmethod
-    def _add_snapshot_intervals(self, snapshot_intervals: SnapshotIntervals) -> None:
+    def add_snapshots_intervals(self, snapshots_intervals: t.Sequence[SnapshotIntervals]) -> None:
```
Comment (izeigerman) [inline sqlmesh/core/state_sync/base.py:402]: @eakmanrq FYI this is the public AP change.

File: sqlmesh/core/state_sync/base.py [hunk 2]
```diff
@@ -450,7 +432,23 @@ def add_interval(
             intervals=intervals if not is_dev else [],
             dev_intervals=intervals if is_dev else [],
         )
-        self._add_snapshot_intervals(snapshot_intervals)
+        self.add_snapshots_intervals([snapshot_intervals])
+
+    @abc.abstractmethod
+    def remove_intervals(
```
Comment (izeigerman) [inline sqlmesh/core/state_sync/base.py:438]: @eakmanrq FYI renamed to `remove_intervals` (plural).

File: sqlmesh/core/state_sync/base.py [hunk 3]
```diff
@@ -312,7 +312,7 @@ def invalidate_environment(self, name: str) -> None:
         """
 
     @abc.abstractmethod
-    def remove_interval(
+    def remove_intervals(
```
Comment (izeigerman) [inline sqlmesh/core/state_sync/base.py:315]: @eakmanrq FYI renamed this to `remove_intervals` (plural)

File: sqlmesh/core/state_sync/engine_adapter.py [hunk 4]
```diff
@@ -221,6 +232,240 @@ def _push_snapshots(self, snapshots: t.Iterable[Snapshot], overwrite: bool = Fal
...
+    def finalize(self, environment: Environment) -> None:
+        """Finalize the target environment, indicating that this environment has been
+        fully promoted and is ready for use.
+
+        Args:
+            environment: The target environment to finalize.
+        """
+        logger.info("Finalizing environment '%s'", environment.name)
+
+        environment_filter = exp.EQ(
```
Comment (tobymao) [inline sqlmesh/core/state_sync/engine_adapter.py:377]: exp.column("name").eq(environment.name)

**Pull request merged**